adding comparison tolerance feature

lewisgross1296 opened the pull request:
This PR is intended to add the comparisonTolerance function to the Utility_QuantityTraits class so that comparisons with a tolerance of 1e-15 can be made. Due to floating point error, some material comparison had a slightly larger error than the previously allowed machine precision. This PR is also made to isolate the smaller problem away from the larger issues in the neutron data reader bug fix PR: #196

gonuke reviewed (approved):
This LGTM, but I'm not an expert on the concept of traits and can't really be sure by just looking at it if it is 100% correct. I'll rely on other reviewers for that.

Is CI functional enough to rely on it yet (@bam241)?

gonuke commented:
Is this waiting for anything?

psbritt commented:
> Is this waiting for anything?

Since this stemmed from the neutron data reader bug which had a lot of unit tests failing, have the tests been run for this PR?

lewisgross1296 commented:
I was waiting for more reviewers before merging. I did not yet run the tests (I will look into this tomorrow and post the results). I separated this from the earlier issue and changes in the Data_XSSNeutronReader.cpp, so it shouldn't be failing those same tests, but tests are still necessary.

bam241 commented:
CI is failing because note setup here.

lewisgross1296 commented:
Rebuilt monte_carlo_core and utility_core packages (only packages changed) and ran all tests for all of them. All 31/31 tests in monte_carlo passed and all 103/103 tests in utility pass. So it should be good for merge. I think it is best practice for someone other than me to merge. @gonuke @psbritt @bam241 @aprobinson

bam241 reviewed (changes requested):
Maybe add test for the new tolerance ?

lewisgross1296 commented:
Good point, I will add the new method for the comparison tolerance to the relevant test and report back when it rebuilds and passes.

lewisgross1296 commented:
This is with respect to the tests I just added. Since we added the `comparisonTolerance()` method for when (machine) epsilon is too strict due to floating point errors, I mimicked the two tests for epsilon in lines 1556 - 1583  of `tstQuantityTraits.cpp` (from the commit by me right before this comment).  I am new to using BOOST's testing framework, but if someone could explain the purpose of each of these 4 check equal tests it would help. 
 
Essentially, I am not super sure the appropriate way to change the last two BOOST_CHECK_EQUAL statements for the new comparisonTolerance feature (lines 1596 - 1612). I am bit stuck on what  `QuantityType::from_value(args)` is supposed to do. I suspect lines 1605 and 1611 are not completely correct, but was hoping for feedback on how to modify them. Maybe some type/type-casting issue.

bam241 commented:
this looks good, have you run the tests locally ?

lewisgross1296 commented:
Yep! Since the only test I changed was in utility_core, I rebuilt the tests with `make -j8 utility_core_tests` and then tested them with `make test-utility_core` and all 103/103 tests passed locally.

psbritt commented:
Many different subpackages throughout FRENSIE rely on the utility subpackages in order to function properly. In general, you need to run the entire unit test suite to ensure your change didn’t break something somewhere else.

> On Jan 13, 2021, at 7:23 PM, Lewis Gross <notifications@github.com> wrote:
> 
> ﻿
> Yep! Since the only test I changed was in utility_core, I rebuilt the tests with make -j8 utility_core_tests and then tested them with make test-utility_core and all 103/103 tests passed locally.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

lewisgross1296 commented:
@psbritt did you mean run all utility package tests or all FRENSIE test? To be safe, I rebuilt FRENSIE and ran all tests and everything passed, except for one test that I think we had previously determined to be okay to ignore. Test 504 - ElectroionizationSubshellElectroatomicReactionNative_test (Failed), but I think this is due to outdated, hard-coded comparison of cross-sections and maybe we should open an issue to resolve this test. Though it seems low priority for anyone not doing electron transport.

psbritt commented:
Yes, I meant all FRENSIE tests. Any time I make a change to any of the source code, I make sure to run all FRENSIE unit tests to ensure that my change didn't cause something else to break that I wasn't aware of. I vaguely remember that test being an issue also - if it is still failing, then an issue should probably be made of it so that documentation of it is made and we're aware of it in the future.

bam241 commented:
Thinking about it, I think @psbritt request to run all the tests would make sense :)

Implementation is good for me tho, thanks for adding this in @lewisgross1296

lewisgross1296 commented:
If it looks good to everyone, I will merge after @psbritt and @bam241 change their status to approve.

bam241 commented:
I am confused here, have you build and test all the FRENSIE packages ?

lewisgross1296 commented:
Yep! I rebuilt and reran all of FRENSIE's tests and they all passed except for the one that I mentioned above.

bam241 commented:
sounds good to me, thx @lewisgross1296 
Merging!
